numpy/core/src/common/simd/vsx/arithmetic.h

```diff
@@ -118,6 +118,28 @@
 
 // Horizontal add: Calculates the sum of all vector elements.
 
+NPY_FINLINE npy_uint16 npyv_sumup_u8(npyv_u8 a)
+{
+    const npyv_u32 zero = npyv_zero_u32();
+    npyv_u32 four = vec_sum4s(a, zero);
+    npyv_u32 one  = vec_sums((npyv_s32)sum4, (npyv_s32)zero4);
+    return (npy_uint16)vec_extract(one, 3);
+}
+
+NPY_FINLINE npy_uint32 npyv_sumup_u16(npyv_u16 a)
+{
+    const npyv_s32 zero = npyv_zero_s32();
+    npyv_u32x2 eight = npyv_expand_u32_u16(a);
+    npyv_u32   four  = vec_add(eight.val[0], eight.val[1]);
+    npyv_s32   one   = vec_sums((npyv_s32)four, zero);
```

Comment:
How does this prevent errors when the cast overflows?

Reply:
there's no overflow since summation not going to exceed the signed int limit. `0x7FFFFFFF-0xFFFF*8 > 0`